### PR TITLE
net/usrsock: add support for CONFIG_NET_ALLOC_CONNS

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/usrsock: add support for CONFIG_NET_ALLOC_CONNS

## Impact

N/A

Depends on:
https://github.com/apache/incubator-nuttx/pull/5154
https://github.com/apache/incubator-nuttx-apps/pull/956

## Testing

CI-check

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: net/usrsock/usrsock_conn.c [hunk 1]
```diff
@@ -45,7 +46,9 @@
 
 /* The array containing all usrsock connections. */
 
+#ifndef CONFIG_NET_ALLOC_CONNS
 static struct usrsock_conn_s g_usrsock_connections[CONFIG_NET_USRSOCK_CONNS];
+#endif
```
Comment (pkarashchenko) [inline net/usrsock/usrsock_conn.c:51]: What about
```
int usrsock_connidx(FAR struct usrsock_conn_s *conn)
{
  int idx = conn - g_usrsock_connections;

  DEBUGASSERT(idx >= 0);
  DEBUGASSERT(idx < ARRAY_SIZE(g_usrsock_connections));

  return idx;
}
```
???

File: net/usrsock/usrsock_conn.c [hunk 2]
```diff
@@ -94,23 +97,36 @@ static void _usrsock_semgive(FAR sem_t *sem)
...
+      conn = kmm_zalloc(sizeof(*conn) * CONFIG_NET_USRSOCK_CONNS);
+      if (conn != NULL)
+        {
+          for (i = 0; i < CONFIG_NET_USRSOCK_CONNS; i++)
+            {
+              dq_addlast(&conn[i].node, &g_free_usrsock_connections);
+            }
+        }
+    }
+#endif
```
Comment (pkarashchenko) [inline net/usrsock/usrsock_conn.c:119]: Why not to do allocation in `usrsock_initialize`?

File: net/usrsock/usrsock_conn.c (hunk 1 above)
Comment (anchao) [reply net/usrsock/usrsock_conn.c:51]: this function will be removed from #5154

File: net/usrsock/usrsock_conn.c (hunk 2 above)
Comment (anchao) [reply net/usrsock/usrsock_conn.c:119]: allocate on use, all conn instances will be obtained through the allocator if g_free_usrsock_connections is NULL

Comment (pkarashchenko) [reply net/usrsock/usrsock_conn.c:119]: In such case we do not limit the allocated connection number by `CONFIG_NET_ALLOC_CONNS`. The current condition seems to be true for the first time and when `CONFIG_NET_ALLOC_CONNS` number of connections are allocated.

Comment (pkarashchenko) [reply net/usrsock/usrsock_conn.c:119]: So why we are not allocating one by one, but with `CONFIG_NET_ALLOC_CONNS` portions?

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**